### PR TITLE
Cam pipe and halide harris merge

### PR DESCRIPTION
**Pull request by Bastacyclop**
Includes #44 and #52.

### CODE REVIEW

File: src/main/scala/shine/C/TranslationContext.scala [hunk 1]
```diff
@@ -15,9 +15,12 @@ class TranslationContext() extends shine.DPIA.Compilation.TranslationContext {
...
 
-      //TODO think about this more thoroughly
-      case PairType(_, _) => Assign(dt, lhs, rhs)
-        // assign(dt1, recordAcc1(dt1, dt2, lhs), fst(rhs)) `;` assign(dt2, recordAcc2(dt1, dt2, lhs), snd(rhs))
+      // FIXME: both solutions currently create issues
+      // TODO: think about this more thoroughly
+      // case PairType(_, _) => Assign(dt, lhs, rhs)
+      case PairType(dt1, dt2) =>
+        assign(dt1, pairAcc1(dt1, dt2, lhs), fst(rhs)) `;`
+        assign(dt2, pairAcc2(dt1, dt2, lhs), snd(rhs))
```
Comment (Bastacyclop) [inline src/main/scala/shine/C/TranslationContext.scala:23]: I plan to open an issue about this part.

Comment (michel-steuwer) [reply src/main/scala/shine/C/TranslationContext.scala:23]: yes, please.

File: src/main/scala/shine/OpenCL/FlagPrivateArrayLoops.scala [hunk 2]
```diff
@@ -28,6 +28,9 @@ object FlagPrivateArrayLoops {
       extends VisitAndRebuild.Visitor
     {
       override def phrase[T <: PhraseType](p: Phrase[T]): Result[Phrase[T]] = p match {
+        // FIXME: we would like this primitive to be eliminated before this pass?
```
Comment (michel-steuwer) [inline src/main/scala/shine/OpenCL/FlagPrivateArrayLoops.scala:31]: Can you also create an issue for this one?

File: src/main/scala/util/package.scala [hunk 3]
```diff
@@ -57,4 +65,25 @@ package object util {
...
+  ): Unit = r match {
+    case elevate.core.Success(p) => dotPrintTmp(name, p)
+    case _ =>
+  }
+
+  def dotPrintTmp(name: String, e: rise.core.Expr): Unit = {
+    import scala.language.postfixOps
+    import scala.sys.process._
+
+    val path = "/tmp"
```
Comment (michel-steuwer) [inline src/main/scala/util/package.scala:81]: Why don't you use the `writeToTempFile` function that uses the `createTempFile` function and avoids hardcoding the `/tmp` path (that e.g. will not work on Windows).

File: src/test/scala/apps/KMeans.scala [hunk 4]
```diff
@@ -8,7 +8,8 @@ class KMeans extends shine.test_util.TestsWithExecutor {
   private val C = 5
   private val F = 34
 
-  test("kmeans versions produce same results") {
+  // FIXME: got broken
+  ignore("kmeans versions produce same results") {
```
Comment (michel-steuwer) [inline src/test/scala/apps/KMeans.scala:12]: Can we create an issue for this as well (and also what does `got broken` mean? does it no longer type check, compiler, pass the test?)

File: src/test/scala/apps/NBody.scala [hunk 5]
```diff
@@ -26,7 +26,8 @@ class NBody extends shine.test_util.TestsWithExecutor {
     ))
   }
 
-  test("nbody AMD version calls update only once") {
+  // FIXME: got broken
+  ignore("nbody AMD version calls update only once") {
```
Comment (michel-steuwer) [inline src/test/scala/apps/NBody.scala:30]: Same here?

**Review by michel-steuwer — APPROVED** [review]
Could you please comment on the individual points I raised?
I am happy with the overall PR and will therefore already approve it, but I still expect your comments.

File: src/test/scala/apps/KMeans.scala (hunk 4 above)
Comment (Bastacyclop) [reply src/test/scala/apps/KMeans.scala:12]: I will write a better comment, but these tests are broken by the pair assignment change.

Comment (michel-steuwer) [reply src/test/scala/apps/KMeans.scala:12]: ok. thanks. Please mention them in the issue that you will create.

File: src/main/scala/shine/C/TranslationContext.scala (hunk 1 above)
Comment (Bastacyclop) [reply src/main/scala/shine/C/TranslationContext.scala:23]: https://github.com/rise-lang/shine/issues/58 is opened

File: src/main/scala/shine/OpenCL/FlagPrivateArrayLoops.scala (hunk 2 above)
Comment (Bastacyclop) [reply src/main/scala/shine/OpenCL/FlagPrivateArrayLoops.scala:31]: I added a comment in https://github.com/rise-lang/shine/issues/22

File: src/main/scala/util/package.scala (hunk 3 above)
Comment (Bastacyclop) [reply src/main/scala/util/package.scala:81]: I can use `File.createTempFile` but it should not be deleted on exit, since this is used for debug.

Comment (Bastacyclop) [reply src/main/scala/util/package.scala:81]: Maybe it is better if I provide a path parameter and set `/tmp` as the default?

Comment (michel-steuwer) [reply src/main/scala/util/package.scala:81]: Well than use directly the `File.createTempFile(prefix, suffix)` method from Java.io.File without calling the `deleteOnExit` method.

**Review by michel-steuwer — DISMISSED** [review]

**Review by michel-steuwer — APPROVED** [review]

**Pull request merged**